### ISSUE TITLE
Strip leading `|` in match arm patterns

### ISSUE THREAD

**Issue by nrc** [rust-lang-nursery/rustfmt#2621]
See discussion in https://github.com/rust-lang-nursery/fmt-rfcs/issues/119. We might need to revisit this if the discussion changes direction.

**Comment by topecongiro** [comment]
Should we have an option for preserving/removing the leading `|`?

**Comment by nrc** [comment]
I don't think we need an option here. I think we probably want to support #2234 as an option for those who want a leading `|`.

**Comment by nrc** [comment]
@Mike-Baker awesome! You should look at https://github.com/rust-lang-nursery/rustfmt/blob/master/src/matches.rs in particular `rewrite_match_arms`. You probably want to look at https://github.com/rust-lang/rust/blob/master/src/libsyntax/ast.rs#L896-L901 too which is where match arms are defined

**Comment by nrc** [comment]
Current master should be building now, please rebase.

The branch looks good, I don't think you should worry abou reverting ea3c01e - #2234 is a bit different to that in that it would be an option to *always* insert a leading `|`, not preserving the way `|` was used in the source. So I think landing your branch is a better plan than looking at #2234 (though that would be a great follow-up issue, if you are interested). 

For the tests, I think `match_with_beginning_vert` needs renaming. You should have tests for no `|` at all, trailing `|` kept if necessary (I'm sure this alreay exists), leading `|` stripped, and maybe some variations with newlines/comments if you can think of some.

**Comment by nrc** [comment]
@Mike-Baker Hey, did you get a chance to rebase your branch? Any more questions I can help with?

**Comment by Mike-Baker** [comment]
@nrc Hi, sorry for the delay.

Previous weekends I have been running into errors

```error[E0308]: mismatched types
    --> /Users/Mike/.cargo/registry/src/github.com-1ecc6299db9ec823/rustc-ap-rustc_errors-113.0.0/emitter.rs:1393:39
     |
1393 |     (a_start..a_end + extra).contains(&b_start)
     |                                       ^^^^^^^^
     |                                       |
     |                                       expected usize, found &usize
     |                                       help: consider removing the borrow: `b_start`
     |
     = note: expected type `usize`
                found type `&usize`

error: aborting due to 2 previous errors
```
and
```
error[E0599]: no method named `end` found for type `std::ops::RangeInclusive<u128>` in the current scope
   --> /Users/Mike/.cargo/registry/src/github.com-1ecc6299db9ec823/rustc-ap-rustc_target-121.0.0/abi/mod.rs:559:37
    |
559 |         let end = *self.valid_range.end();
    |                                     ^^^ field, not a method
    |
    = help: did you mean to write `self.valid_range.end` instead of `self.valid_range.end(...)`?
    = help: items from traits can only be used if the trait is in scope
    = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
            candidate #1: `use std::ops::RangeBounds;`

error: aborting due to 2 previous errors
```

The most recent (still in effect)
```
error: process didn't exit successfully: `/Users/Mike/sync/programming/rust/rustfmt/target/debug/deps/rustfmt_nightly-ed88e69f497ca24a` (signal: 11, SIGSEGV: invalid memory reference)
```
On OSX with ```
cargo 1.28.0-nightly (f352115d5 2018-05-15)
rustc 1.28.0-nightly (952f344cd 2018-05-18)
commit e83c7ff0056d522a722e26c24e7707f5d84adba4
Merge: dab99b6 912e4bd
Author: Nick Cameron <nrc@ncameron.org>
Date:   2018-05-19 11:47:36 +1200

    Merge pull request #2723 from topecongiro/rustc-ap-syntax
    
    Update rustc-ap-syntax

```

**Comment by nrc** [comment]
@Mike-Baker you'll need to `rustup update` and rebase to current master so that your compiler libs match the rustfmt libs. I think that should fix these issues, let me know if it doesn't.

**Comment by nrc** [comment]
@Mike-Baker did you get a chance to rebase and did that solve your problems? Anything else I can help with?

**Comment by Mike-Baker** [comment]
@nrc Sorry for the delay, I have finally got everything building again so I have rebased and got the tests in line.

Here are the [changes](https://github.com/Mike-Baker/rustfmt/commit/2dae1eb417e3abb8175dada96c0fe63b92e66371) since last time.

Here is the [pull request](https://github.com/rust-lang-nursery/rustfmt/pull/2804)

**Comment by Mike-Baker** [comment]
Minor notes: It seems that things don't build on current nightly (fails with ```error: failed to run custom build command for `rustc-ap-rustc_target v164.0.0`
process didn't exit successfully: `./rustfmt/target/debug/build/rustc-ap-rustc_target-7693b08837a4e8f0/build-script-build` (signal: 11, SIGSEGV: invalid memory reference)```)

But it does build on nightly-2018-06-20.

I also managed to get to a state somewhere that running `cargo test` wasn't correctly replacing the rustfmt executable which I fixed by nuking the `target` directory, I'm sorry to say that if I had figured this out earlier I could have saved a lot of time.

**Comment by nrc** [comment]
Closed by #2804
